### PR TITLE
[sil] Refactor the open_existential_addr immutable use checking into …

### PR DESCRIPTION
**Pull request by gottesmm**
…its own verifier utility.

My intention is to use this checker to also verify that in_guaranteed arguments
are used in the same manner immutably. Beyond improving that in_guaranteed
parameters are properly used immutably (which is just goodness), by using the
same check it ensures that we can always inline a callee into a caller with an
open_existential_addr without violating any check on the oea instruction.

I am doing this separately from adding more checks/applying it to
in_guaranteed/fixing some exposed bugs for ease of review. This change is
completely mechanical.

rdar://50212579

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci smoke test and merge

File: lib/SIL/SILVerifier.cpp [hunk 1]
```diff
@@ -410,6 +410,130 @@ void verifyKeyPathComponent(SILModule &M,
   baseTy = componentTy;
 }
 
+/// Check if according to the SIL language model this memory /must only/ be used
+/// immutably. Today this is only applied to in_guaranteed arguments and
+/// open_existential_addr. We should expand it as needed.
+struct ImmutableAddressUseVerifier {
+  SmallVector<Operand *, 32> worklist;
```
Comment (gottesmm) [inline lib/SIL/SILVerifier.cpp]: NOTE: I am going to use this in a forthcoming commit to look through certain instructions recursively.

**Comment by gottesmm** [comment]
@swift-ci smoke test os x platform

**Comment by rxwei** [comment]
Very cool! The new verifier helped me find a terrible bug in the automatic differentiation code path.

**Comment by gottesmm** [comment]
@rxwei Nice! We need more of these static checks in general IMO. Makes it so much easier to debug these sorts of things. If you have ideas for more, I am all ears. I can also imagine an analysis that ensures that values are always initialized before they are used /throughout/ the compilation pipeline. This would embed the semantics. NOTE This is strictly different than DI since DI is checking unchecked user input. This is about defining the semantics at the SIL level.

**Pull request merged**